File: Pod/Classes/ABEReporterViewController.m
//
//  ABEReporterViewController.m
//  IssueReporter
//
//  Created by Nikolai Johan Heum on 04.05.15.
//  Copyright (c) 2015 Abello. All rights reserved.
//

// View Controllers
#import "ABEReporterViewController.h"

// Model
#import "ABEIssue.h"

// Helpers
#import "ABEGithubAPIClient.h"
#import "ABEImgurAPIClient.h"
#import "ABEReporter.h"
#import "ABEIssueManager.h"

// Categories
#import "UIBarButtonItem+ABECustomBarButton.h"
#import "UIColor+ABEColor.h"
#import "NSFileManager+ABEFileManager.h"
#import "NSBundle+ABEBundle.h"
#import "UIAlertController+ABEErrorAlertController.h"
#import "NSThread+ABEMainThreadRunner.h"

static NSString * const kABETableName = @"IssueReporter-Localizable";

static NSString * const kABEPlaceHolderString = @"What went wrong?";
static NSString * const kABETitle =             @"New issue";
static NSString * const kSaveImageName =        @"save";

static int const kABEdescriptionTextViewCornerRadius   = 4;
static double const kABEdescriptionTextViewBorderWidth = 0.5;

static CGFloat const kABETextFieldInset = 14;

static void *ABEImageUploadCountObservingContext = &ABEImageUploadCountObservingContext;

@interface ABEReporterViewController () <UITextFieldDelegate>

@property (nonatomic, weak) IBOutlet UITextView *descriptionTextView;
@property (nonatomic, weak) IBOutlet UITextField *titleTextField;
@property (nonatomic, weak) IBOutlet UILabel *placeHolderLabel;

@property (nonatomic, weak) ABEImageCollectionViewController *imageCollectionViewController;
@property (nonatomic) ABEIssueManager *issueManager;
@end

@implementation ABEReporterViewController

+ (instancetype)instance {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:NSStringFromClass([self class]) bundle:[NSBundle abe_bundleForLibrary]];
    ABEReporterViewController *reporterViewController = [storyboard instantiateInitialViewController];
    return reporterViewController;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        self.title = kABETitle;
    }
    
    return self;
}

- (UIView *)snapshotCurrentViewState {
    return [UIApplication sharedApplication].delegate.window;
}

- (void)loadView {
    [super loadView];
    
    self.issueManager = [[ABEIssueManager alloc] initWithReferenceView:[self snapshotCurrentViewState] viewController:self];
}

- (void)viewDidLoad {
    [super viewDidLoad];

    [self configureBlueBar];
    [self configureTextViews];
    [self setupLocalization];
    
    [self.titleTextField addTarget:self action:@selector(textFieldDidChange:) forControlEvents:UIControlEventEditingChanged];
    
    self.navigationItem.leftBarButtonItem = [UIBarButtonItem njh_backButtonWithTarget:self andColor:[UIColor whiteColor] action:@selector(dismissIssueReporter)];
    self.navigationItem.rightBarButtonItem = [UIBarButtonItem njh_saveButtonWithTarget:self action:@selector(saveIssue)];
    
    [self.issueManager addObserver:self forKeyPath:NSStringFromSelector(@selector(imagesToUpload)) options:NSKeyValueObservingOptionInitial context:ABEImageUploadCountObservingContext];
}

- (void)dealloc {
    [self.issueManager removeObserver:self forKeyPath:NSStringFromSelector(@selector(imagesToUpload)) context:ABEImageUploadCountObservingContext];
}

/**
 *  Adds a spacer on the left side of the top text field and adds a border around the text view
 */
- (void)configureTextViews {
    UIView *spacerView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, kABETextFieldInset, kABETextFieldInset)];
    self.titleTextField.leftViewMode = UITextFieldViewModeAlways;
    self.titleTextField.leftView = spacerView;
    
    self.descriptionTextView.layer.borderColor = [UIColor njh_greyBorderColor].CGColor;
    self.descriptionTextView.layer.cornerRadius = kABEdescriptionTextViewCornerRadius;
    self.descriptionTextView.layer.borderWidth = kABEdescriptionTextViewBorderWidth;
    self.descriptionTextView.textContainerInset = UIEdgeInsetsMake(kABETextFieldInset, kABETextFieldInset, 0, kABETextFieldInset);
}

- (void)configureBlueBar {
    self.navigationController.navigationBar.barTintColor = [UIColor njh_blueNavigationBarColor];
    [self.navigationController.navigationBar setTitleTextAttributes:@{NSForegroundColorAttributeName : [UIColor whiteColor]}];
}

- (void)setupLocalization {
    NSBundle *bundle = [NSBundle bundleForClass:self.class];
    
    self.titleTextField.text = NSLocalizedStringFromTableInBundle(self.titleTextField.text, kABETableName, bundle, nil);
    self.placeHolderLabel.text = NSLocalizedStringFromTableInBundle(self.placeHolderLabel.text, kABETableName, bundle, nil);
    self.title = NSLocalizedStringFromTableInBundle(self.title, kABETableName, bundle, nil);
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSString *,id> *)change context:(void *)context {
    if (context == ABEImageUploadCountObservingContext) {
        [NSThread abe_guaranteeBlockExecutionOnMainThread:^{
            if (self.issueManager.imagesToUpload.count == 0) {
                self.navigationItem.rightBarButtonItem = [UIBarButtonItem njh_saveButtonWithTarget:self action:@selector(saveIssue)];
                self.navigationItem.rightBarButtonItem.enabled = YES;
            } else {
                UIActivityIndicatorView *spinner = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhite];
                [spinner startAnimating];
                
                self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:spinner];
                self.navigationItem.rightBarButtonItem.enabled = NO;
            }
        }];
    }
}

- (void)dismissIssueReporter {
    [NSFileManager njh_clearDocumentsDirectory];
    
    [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - Actions

- (void)saveIssue {
    __weak typeof(self) _self_weak = self;
    [self.issueManager saveIssueWithCompletion:^{
        __strong typeof(self) self = _self_weak;
        [self userDidTryToHideKeyboard:nil]; // animate the keyboard away
        [self dismissIssueReporter];
    }];
}

- (IBAction)userDidTryToHideKeyboard:(id)sender {
    [self.view endEditing:false];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"embed_segue"]) {
        self.imageCollectionViewController = segue.destinationViewController;
        self.imageCollectionViewController.issueManager = self.issueManager;
    }
}

- (void)textFieldDidChange:(UITextField *)textField {
    if (textField == self.titleTextField) {
        self.issueManager.issue.title = textField.text;
    }
}

- (void)textViewDidChange:(UITextView *)textView {
    if (textView == self.descriptionTextView) {
        self.placeHolderLabel.hidden = ![textView.text isEqualToString:@""];
        self.issueManager.issue.issueDescription = textView.text;
    }
}

@end

File: Pod/Classes/ABEReporterViewController.swift
//
//  ABEReporterViewController.swift
//  Pods
//
//  Created by Hakon Hanesand on 10/9/16.
//
//

import Foundation
import UIKit
import CoreGraphics

class ABEReporterViewController: UIViewController {
    
    private static let kABETextFieldInset = 14
    
    private static let kABEdescriptionTextViewCornerRadius = CGFloat(4);
    private static let kABEdescriptionTextViewBorderWidth = CGFloat(0.5);
    
    private static let kABETableName = "IssueReporter-Localizable"
    
    @IBOutlet weak var descriptionTextView: UITextView!
    @IBOutlet weak var titleTextField: UITextField!
    @IBOutlet weak var placeHolderLabel: UILabel!
    
    fileprivate var imageCollectionViewController: ABEImageCollectionViewController!
    
    public var issueManager: ABEIssueManager! {
        didSet {
            issueManager.delegate = self
        }
    }
    
    public class func instance(withIssueManager manager: ABEIssueManager) -> ABEReporterViewController {
        let storyboard = UIStoryboard(name: String(describing: self), bundle: Bundle.bundleForLibrary())
        let reporterViewController = storyboard.instantiateInitialViewController() as! ABEReporterViewController
        
        reporterViewController.issueManager = manager
        
        return reporterViewController
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        configureTextView()
        setupLocalization()
        
        navigationController?.navigationBar.barTintColor = UIColor.blueNavigationBarColor()
        navigationController?.navigationBar.titleTextAttributes = [NSForegroundColorAttributeName : UIColor.white]
    }
    
    private func configureTextView() {
        
        let spacerView = UIView(frame: CGRect(x: 0, y: 0, width: ABEReporterViewController.kABETextFieldInset, height: ABEReporterViewController.kABETextFieldInset))
        
        titleTextField.leftViewMode = .always
        titleTextField.leftView = spacerView
        
        descriptionTextView.layer.borderColor = UIColor.greyBorderColor().cgColor
        descriptionTextView.layer.cornerRadius = ABEReporterViewController.kABEdescriptionTextViewCornerRadius
        descriptionTextView.layer.borderWidth = ABEReporterViewController.kABEdescriptionTextViewBorderWidth
        
        let textFieldInset = CGFloat(ABEReporterViewController.kABETextFieldInset)
        descriptionTextView.textContainerInset = UIEdgeInsets(top: textFieldInset, left: textFieldInset, bottom: 0, right: textFieldInset)
        descriptionTextView.delegate = self
    }
    
    private func setupLocalization() {
        titleTextField.text = NSLocalizedString(titleTextField.text!, tableName: ABEReporterViewController.kABETableName, bundle: Bundle.bundleForLibrary(), comment: "title of issue")
        placeHolderLabel.text = NSLocalizedString(placeHolderLabel.text!, tableName: ABEReporterViewController.kABETableName, bundle: Bundle.bundleForLibrary(), comment: "placeholder")
        title = NSLocalizedString(self.navigationItem.title!, tableName: ABEReporterViewController.kABETableName, bundle: Bundle.bundleForLibrary(), comment: "title")
    }
    
    @IBAction func cancelIssueReporting(_ sender: AnyObject) {
        self.presentingViewController?.dismiss(animated: true)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let identifier = segue.identifier, identifier == "embed_segue" {
            self.imageCollectionViewController = segue.destination as! ABEImageCollectionViewController
            self.imageCollectionViewController.issueManager = self.issueManager
        }
    }
    
    public func saveIssue() {
        
        issueManager.issue.title = titleTextField.text
        issueManager.issue.issueDescription = descriptionTextView.text

        issueManager.saveIssue { [weak self] in
            DispatchQueue.main.async { [weak self] in
                self?.view.endEditing(false)
                self?.dismissIssueReporter()
            }
        }
    }
    
    fileprivate func dismissIssueReporter() {
        FileManager.clearDocumentsDirectory()
        presentingViewController?.dismiss(animated: true)
    }
    
    deinit {
        DispatchQueue.global(qos: .background).async {
            FileManager.clearDocumentsDirectory()
        }
    }
}

extension ABEReporterViewController: ABEIssueManagerDelegate {

    func issueManagerUploadingStateDidChange(issueManager: ABEIssueManager) {
        DispatchQueue.main.async {
            self.imageCollectionViewController?.collectionView?.reloadData()
            
            if issueManager.isUploading {
                let spinner = UIActivityIndicatorView(activityIndicatorStyle: .white)
                
                self.navigationItem.rightBarButtonItem = UIBarButtonItem(customView: spinner)
                self.navigationItem.rightBarButtonItem?.isEnabled = false
                
                spinner.startAnimating()
                
            } else {
                self.navigationItem.rightBarButtonItem = UIBarButtonItem.saveButton(self, action: #selector(ABEReporterViewController.saveIssue))
                self.navigationItem.rightBarButtonItem?.isEnabled = true
            }
        }
    }
    
    func issueManager(_ issueManager: ABEIssueManager, didFailToUploadIssueWithError error: Error) {
        let alertController = UIAlertController(error: error as NSError)
        
        alertController.addAction(UIAlertAction(title: "Retry", style: .default) { [weak self] action in
            issueManager.saveIssue { [weak self] in
                DispatchQueue.main.async { [weak self] in
                    self?.view.endEditing(false)
                    self?.dismissIssueReporter()
                }
            }
        })
    }
}

extension ABEReporterViewController: UITextViewDelegate {
    
    func textViewDidChange(_ textView: UITextView) {
        let length = textView.text?.characters.count ?? 0
        placeHolderLabel.isHidden = length > 0
    }
}
